cmd/checksum.go

```diff
@@ -0,0 +1,39 @@
+package cmd
+
+import (
+	"fmt"
+
+	"github.com/jpbetz/auger/pkg/data"
+	"github.com/spf13/cobra"
+)
+
+var checksumCmd = &cobra.Command{
+	Use:   "checksum",
+	Short: "Checksum a etcd keyspace.",
+	RunE: func(cmd *cobra.Command, args []string) error {
+		return checksum()
+	},
+}
+
+type checksumOptions struct {
+	filename string
+	revision int64
+}
+
+var checksumOpts *checksumOptions = &checksumOptions{}
+
+func init() {
+	RootCmd.AddCommand(checksumCmd)
+	checksumCmd.Flags().StringVarP(&checksumOpts.filename, "file", "f", "", "Bolt DB '.db' filename")
+	checksumCmd.Flags().Int64VarP(&checksumOpts.revision, "revision", "r", 0, "etcd revision to retrieve data at, if 0, the latest revision is used, defaults to 0")
```

Comment:
We know that each boltdb file will have a range of uncompacted revisions. In the HA case these are unlikely to line up well. So if we just print the latest, then you may need to run with the 0 options 1 or 2 additional times to find a revision which is present everywhere. Might it make sense to just print all uncompacted revisions (or have an option to print all revisions)? If you find a discrepancy it seems like something you will want to do to see if the corruption event occurred in you compaction window anyway.

Reply:
Since revisions are monotonically increasing, I'll add the compaction revision so the range is known. That seems handy and should make revision selection across members trivial.  And if one of them reports a range way out of wack with the others, then it makes it really obvious what the drift is between them.